Fix bug with fetching projects from SSH URIs

smurching opened the pull request:
Fixes a bug introduced in #215 where we'd fetch project URIs of the form `"git@github.com:/some/path"` into a local directory with name `"git@github.com:/some/path"` instead of into a temporary directory, and adds a regression test.

smurching committed:
Bugfix for fetching projects from SSH uris

smurching commented on tests/projects/test_projects.py:
```diff
@@ -59,11 +59,12 @@ def test_fetch_project(local_git_repo, local_git_repo_uri):
     for force_tempdir, uri in [(None, TEST_PROJECT_DIR), (False, TEST_PROJECT_DIR)]:
         assert mlflow.projects._fetch_project(uri=uri, force_tempdir=force_tempdir) == \
                os.path.abspath(TEST_PROJECT_DIR)
+    assert mlflow.projects._should_use_temp_dir("git@localhost:%s.git" % local_git_repo)
```
Note: I didn't want to add a test that invokes `mlflow run git@github.com:databricks/mlflow-example.git` since doing so would add additional dependencies on network / slow the build, but open to doing so. Such a test would also have not caught the bug addressed by this PR (i.e. we'd successfully fetch the project into a local directory named "git@github.com:databricks/mlflow-example.git" the first time we invoked `mlflow run`, and then fail on successive invocations).

Also: it's almost-possible to test fetching an SSH git URI by calling `_fetch_project` with a URI like "sid@localhost:/path/to/local/repo", but doing so requires a password so we perform a less direct test here.

aarondav replied:
We should have a place for integration tests like that, though, as it is the only way to verify the end-to-end functionality. We do have some tests marked "large" today which make network calls, which is our current solution.

The test would just have to call `mlflow run` twice to satisfy your concern. We should also test for both the `git@` and `https://` forms of the Git URLs. I think we should just ahead and add the integration test rather than add more hacky "test-random-internal-methods" sorts of tests.

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -176,13 +177,20 @@ def _get_storage_dir(storage_dir):
+    """
+    Returns True if we should fetch the project at the passed-in URI into a temporary directory.
+    """
+    return _is_git_uri(parsed_uri) or mlflow.data.is_uri(parsed_uri)
```
I don't think it's worth relying on an external implementation for this:

```
    from six.moves import urllib
    parsed_uri = urllib.parse.urlparse(string)
    return len(parsed_uri.scheme) > 0
```

Much harder to understand what it's doing.

smurching committed:
Address review comments, add integration test

smurching commented on mlflow/projects/__init__.py:
```diff
@@ -193,12 +189,12 @@ def _fetch_project(uri, force_tempdir, version=None, git_username=None, git_pass
+    use_temp_dst_dir = force_tempdir or _GIT_URI_REGEX.match(parsed_uri)
+    dst_dir = tempfile.mkdtemp() if use_temp_dst_dir else parsed_uri
     if use_temp_dst_dir:
         eprint("=== Fetching project from %s into %s ===" % (uri, dst_dir))
```
This logic is bug-prone in general in that we compute the `dst_dir` upfront and then use it across all project types (Git & local project). However, it does allow us to log the `dst_dir` before we attempt to fetch the project. One fix would be to allow individual project types to decide whether or not to use a tempdir (for clarity) at the cost of duplicating this logging message

smurching commented on tests/projects/test_projects_cli.py:
```diff
@@ -1,25 +1,29 @@
-from mlflow.utils.file_utils import TempDir
-from tests.integration.utils import invoke_cli_runner, update_temp_env
-from tests.projects.utils import TEST_PROJECT_DIR, GIT_PROJECT_URI
+from tests.integration.utils import invoke_cli_runner
```
Refactoring the existing tests to use mock helpers from elsewhere + adding new integration tests for SSH URIs

aarondav commented on tests/projects/test_projects_cli.py:
```diff
@@ -1,25 +1,29 @@
+def test_run_git_https(tracking_uri_mock):  # pylint: disable=unused-argument
+    # Invoke command twice to ensure we set Git state in an isolated manner (e.g. don't attempt to
+    # create a git repo in the same directory twice, etc)
+    invoke_cli_runner(cli.run, [GIT_PROJECT_URI, "-P", "alpha=0.5"])
```
nit: could we assert that `GIT_PROJECT_URI` startswith `https`? The name of the parameter does not indicate that it's necessarily https, so I could see a future PR alter it without thinking about this particular test. Similar thing with the git test.

smurching committed:
Re-introduce _is_local_uri helper, use it to determine whether to use a tempdir to fetch a project.
Also disable host key checking in integration test that fetches a project via an SSH URI

smurching committed:
Add comment to _is_local_uri

smurching committed:
Add asserts on URIs

smurching committed:
Fix typo in assert on SSH URI

smurching committed:
Mark SSH test as optional

smurching commented:
`test_model_export` failed while attempting to `pip install pyspark` in a docker container, retriggering the build
